nemoguardrails/embeddings/cache.py

```diff
@@ -0,0 +1,195 @@
+
+    def _get_file_path(self, key):
+        return self._cache_dir / str(key)
+
+    def get(self, key):
+        file_path = self._get_file_path(key)
+        if file_path.exists():
+            with open(file_path, "r") as file:
+                return json.load(file)
+        return None
+
+    def set(self, key, value):
+        file_path = self._get_file_path(key)
+        with open(file_path, "w") as file:
+            json.dump(value, file)
+
+
+class RedisCacheStore(CacheStore):
+    """Redis cache store.
+
+    This cache store keeps the cache in a Redis database. It can be used to share the cache between multiple machines.
+
+    Args:
+        redis_client (redis.Redis, optional): The Redis client to use. If not provided, a new client will be created.
+
+    Example:
+        >>> redis_client = redis.Redis(host='localhost', port=6379, db=0)
+        >>> cache_store = RedisCacheStore(redis_client=redis_client)
+        >>> cache_store.set('key', 'value')
+        >>> print(cache_store.get('key'))
+        value
+    """
+
+    def __init__(self, host: str = "localhost", port: int = 6379, db: int = 0):
+        import redis
+
+        self._redis = redis.Redis(host=host, port=port, db=db)
+
+    def get(self, key):
+        return self._redis.get(key)
+
+    def set(self, key, value):
+        self._redis.set(key, value)
+
+
+def cache_embeddings(func):
+    """
+    Decorator to cache the embeddings.
+
+    This decorator caches the embeddings in the cache store.
+    It uses the key generator to generate the key for the cache.
+
+    If the class does not have a `key_generator` attribute, it will use the `MD5KeyGenerator`.
+    If the class does not have a `cache_store` attribute, it will use the `FilesystemCacheStore`.
+
+    This decorator can be applied to any function that accepts a list of strings and returns a list of lists of floats.
+
+    Args:
+        func (Callable[[List[str]], List[List[float]]]): The function to decorate.
+
+    Returns:
+        Callable[[List[str]], List[List[float]]]: The decorated function.
+
+    Example:
+        @cache_embeddings
+        def get_embeddings(texts: List[str]) -> List[List[float]]:
+            # implementation here
+            pass
+    """
+
+    @functools.wraps(func)
+    def wrapper_decorator(self, texts):
+        results = []
+
+        if not hasattr(self, "key_generator"):
+            self.key_generator = MD5KeyGenerator()
+        if not hasattr(self, "cache_store"):
+            self.cache_store = FilesystemCacheStore()
+
+        for text in texts:
+            key = self.key_generator.generate_key(text)
+            result = self.cache_store.get(key)
+            if result is None:
+                print("Does not exist in the cache")
+                result = func(self, text)
+                self.cache_store.set(key, result)
+            else:
+                print("Fetching from the cache")
+            results.append(result)
```

Comment:
This will break the "bulk computing behavior" which most embedding providers should support (e.g. OpenAI API). The logic here should check which ones are in the cache, and only make a single call with the rest. However, since this is done only once in the beginning, we could also just go with a simpler logic: if not all of the texts are found in cache, compute all again and update the cache.

Reply:
Thanks for pointing it out. I implement this logic.

"should check which ones are in the cache, and only make a single call with the rest."